recirq/quantum_chess/initial_mapping_utils.py

```diff
@@ -0,0 +1,383 @@
+    circuit: cirq.Circuit,
+) -> Dict[cirq.Qid, List[Tuple[cirq.Qid, int]]]:
+    """Returns an adjacency graph of logical qubits of the circuit.
+
+    Uses the heuristic of adding an edge between the nodes of each
+    disjoint component that are least connected if the graph contains more
+    than one connected component.
+
+    Each edge is a tuple containing an adjacent node and the index of the
+    moment at which the operation occurs.
+
+    Arg:
+      circuit: The circuit from which to build a logical qubits graph.
+    """
+    g = defaultdict(list)
+    moment_index = 0
+
+    # Build an adjacency graph based on the circuit.
+    for i, m in enumerate(circuit):
+        moment_index = i
+        for op in m:
+            if isinstance(op.gate, cirq.MeasurementGate):
+                # Skip measurement gates.
+                continue
+            if len(op.qubits) == 1:
+                q = op.qubits[0]
+                if q not in g:
+                    g[q] = []
+            elif len(op.qubits) == 2:
+                q1, q2 = op.qubits
+                q1_neighbors = [n[0] for n in g[q1]]
+                if q2 not in q1_neighbors:
+                    g[q1].append((q2, i))
+                q2_neighbors = [n[0] for n in g[q2]]
+                if q1 not in q2_neighbors:
+                    g[q2].append((q1, i))
+            else:
+                raise ValueError(f'Operation {op} has more than 2 qubits!')
+
+    # Find the connected components in the graph.
+    components = deque()
+    visited = set()
+    for q in g:
+        if q not in visited:
+            components.append(traverse(g, q, visited))
+
+    if len(components) == 1:
+        return g
+
+    # Connect disjoint components by adding an edge between the nodes of
+    # each disjoint component that are least connected.
+    while len(components) > 1:
+        moment_index += 1
+        first_comp = components.pop()
+        first_q = get_least_connected_qubit(g, first_comp)
+        second_comp = components.pop()
+        second_q = get_least_connected_qubit(g, second_comp)
+
+        # Add an edge between the two least connected nodes.
+        g[first_q].append((second_q, moment_index))
+        g[second_q].append((first_q, moment_index))
+
+        # Combine the two components and add it back to the components
+        # deque to continue connecting disjoint components.
+        first_comp += second_comp
+        components.append(first_comp)
+
+    return g
+
+
+def find_all_pairs_shortest_paths(
+    g: Union[
+        Dict[cirq.GridQubit, List[cirq.GridQubit]],
+        Dict[cirq.Qid, List[Tuple[cirq.Qid, int]]],
+    ],
+    v: int,
+    m: Dict[cirq.Qid, int],
+) -> List[List[int]]:
+    """ Returns a matrix of the shortest distance between each pair of nodes.
+
+    Implements the Floyd–Warshall algorithm.
+
+    Args:
+      g: A physical qubits graph or a logical qubits graph.
+      v: The size of the graph.
+      m: A mapping of qubit to index.
```

Comment:
would the interface be simpler if we keyed things by Qid directly and returned a dict from Pair[Qid, Qid] -> int instead of requiring an intermediate Qid -> index mapping? Then the only args we'd need to provide would be the graph dict

Reply:
updated and it's much cleaner now!

Reply:
just need to update the return type hint to something like Dict[Tuple[Qid, Qid], int]  (who/what is using these type hints anyways...?)

Reply:
ty for catching that! i'm guessing they're for eventual docs generation...?